File: 127-WordLadder/C++/main.cpp
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        string alphabet = "abcdefghijklmnopqrstuvwxyz";
        set<string> wordSet(wordList.begin(), wordList.end());
        
        if (wordSet.find(endWord) == wordSet.end()) return 0;
        
        size_t size = beginWord.size();
        set<string> curWords{beginWord};
        set<string> endWords{endWord};
        
        set<string> seen; // recall which values we have already seen
        int c = 1; // counter
        
        while (!curWords.empty()) {
            
            // use 2-end BFS
            if (curWords.size() > endWords.size()) {
                set<string> temp(endWords.begin(), endWords.end());
                endWords = curWords;
                curWords = temp;
            }
            
            set<string> nextWords;
            
            for (auto possible_word: curWords) {
                
                // change one letter of each parent word
                for (size_t i = 0; i < size; ++i) {
                    for (auto letter: alphabet) {
                        char prev = possible_word[i];
                        possible_word[i] = letter;
                        
                        // early stopping
                        if (endWords.find(possible_word) != endWords.end()) {
                            return c + 1;
                        }
                        
                        if (seen.find(possible_word) == seen.end() && wordSet.find(possible_word) != wordSet.end()) {
                            seen.insert(possible_word);
                            nextWords.insert(possible_word);
                        }
                        
                        // switch back the letter we modified
                        possible_word[i] = prev;
                    }
                }
            }
            
            // iterate BFS
            curWords = nextWords;
            c++;
            
        }
        
        return 0;        
    }
};

File: 127-WordLadder/Python/main.py
class Solution:    
    def ladderLength(self, beginWord, endWord, wordList):
        """
        :type beginWord: str
        :type endWord: str
        :type wordList: List[str]
        :rtype: int
        """
        alphabet = "abcdefghijklmnopqrstuvwxyz"
        wordSet = set(wordList)

        if endWord not in wordSet:
            return 0

        # we assume all the words have same
        # size so we just need to get the size
        # once and for all
        size = len(beginWord)
        curWords = {beginWord}
        endWords = {endWord}

        # recall which values we already seen
        seen = set()
        c = 1


        while curWords:
            
            # use 2-end BFS. Next expand of the node will
            # be on the set that contains fewer nodes
            if len(curWords) > len(endWords):
                curWords, endWords = endWords, curWords

            nextWords = set()
            for possible_word in curWords:

                # change one letter of each parent word
                for i in range(size):
                    for letter in alphabet:
                        child_word = possible_word[:i] + letter + possible_word[i+1:]

                        # early stopping
                        if child_word in endWords:
                            return c + 1

                        if child_word not in seen and child_word in wordSet:
                            seen.add(child_word)
                            nextWords.add(child_word)

            # iterate BFS
            curWords = nextWords
            c += 1

        return 0